### DISCUSSION TITLE
Properly understanding order of matrices

### DISCUSSION THREAD

**Discussion by SSNTails** [jrouwe/JoltPhysics#998] [Q&A]
I'm coming from Bullet, where I could call .setFromOpenGLMatrix() and .getFromOpenGLMatrix() and receive/pass a float[4][4] and be on my way. This doesn't seem to be as straightforward, or at least, I'm getting inconsistent results and want to ensure I'm properly passing my matrices to Jolt before I ask more questions.

When creating a body, would this be correct?
`

JPH::Mat44 mat = JPH::Mat44(JPH::Vec4(object->transform.m[0][0], object->transform.m[1][0], object->transform.m[2][0], object->transform.m[3][0]),
	JPH::Vec4(object->transform.m[0][1], object->transform.m[1][1], object->transform.m[2][1], object->transform.m[3][1]),
	JPH::Vec4(object->transform.m[0][2], object->transform.m[1][2], object->transform.m[2][2], object->transform.m[3][2]),
	JPH::Vec4(object->transform.m[0][3], object->transform.m[1][3], object->transform.m[2][3], object->transform.m[3][3]))
	JPH::BodyCreationSettings bcs(new JPH::ConvexHullShapeSettings(vertexList), mat.GetTranslation(), mat.GetQuaternion(), JPH::EMotionType::Kinematic, Layers::MOVING);

`

If I don't .Transposed(), I get wildly incorrect results, and if I Transposed() without changing .GetQuaternion() to .GetRotation().GetQuaternion(), I get an assert failure.

But let's start with the correct way to load a [4][4] matrix. Here's what one might look like:

`
+		[0]	0x0000025f3f150ec0 {1.00000000, 0.00000000, 0.00000000, 0.00000000}	float[4]
+		[1]	0x0000025f3f150ed0 {0.00000000, 1.00000000, 0.00000000, 0.00000000}	float[4]
+		[2]	0x0000025f3f150ee0 {0.00000000, 0.00000000, 1.00000000, 0.00000000}	float[4]
+		[3]	0x0000025f3f150ef0 {-146.400009, 14.3999996, 427.200012, 1.00000000}	float[4]
`

**Comment by MrOnlineCoder** [comment]
Can I ask you the purpose of using `setFromOpenGLMatrix` at all? In my understanding, the physics engine and game logic should dictate how and when the position/rotation/scale of object is changed, and then the rendering system just displays it as-is - it's a one way integration.

I usually just have some utility function which can build a model matrix out from JPH:Body, say a glm::mat4:

```cpp
JPH::Body* body = getEntityBody();

glm::mat4 modelMatrix = Utilities::buildModelMatrixForRigidbody(joltBody);

Renderer::draw(someMeshOrModel, modelMatrix);
```

As far as I understand from the code (https://github.com/jrouwe/JoltPhysics/blob/e8a99bfa0c60997eb4a7c22bd5c34f36a87dc632/Jolt/Math/Matrix.h#L228) Jolt also uses column-major matrices, so integrating them with OpenGL or glm should be no problem.

**Reply by SSNTails** [reply]
It's because the starting transformation is loaded from level data, into the physics system. It arrives in the form of float[4][4] in the order exampled above.

**Comment by jrouwe** [comment]
Jolt matrices are column major, I don't know what yours are. So you have ```Mat44(Vec4(axisx, 0), Vec4(axisy, 0), Vec4(axisz, 0), Vec4(translation, 1))```.

Doing `matrix.GetRotation().GetQuaternion()` is the normal way to get the quaternion of a matrix that also happens to contain translation.

**Reply by SSNTails** [reply]
The example I provided is an identity matrix with a XYZ translation of -146.400009, 14.3999996, 427.200012 . I was always under the impression that it was column-major, since I used to pass these directly into OpenGL 1.1 a LONG time ago, but perhaps I'm mistaken?

**Comment by SSNTails** [comment]
I think I've got this figured out -- the proper order for a float[4][4] should be:

`
JPH::Mat44 mat = JPH::Mat44(JPH::Vec4(object->transform.m[0][0], object->transform.m[0][1], object->transform.m[0][2], object->transform.m[0][3]),
	JPH::Vec4(object->transform.m[1][0], object->transform.m[1][1], object->transform.m[1][2], object->transform.m[1][3]),
	JPH::Vec4(object->transform.m[2][0], object->transform.m[2][1], object->transform.m[2][2], object->transform.m[2][3]),
	JPH::Vec4(object->transform.m[3][0], object->transform.m[3][1], object->transform.m[3][2], object->transform.m[3][3]));
`

Thank you for your input on .GetRotation().GetQuaternion(). I am not strong on quaternions, but I understand their usefulness.

Also, the anomalies I was experiencing (which I haven't mentioned yet) were due to convex hulls having a different center of mass than static triangle meshes.
